### PR TITLE
Covariance for complex numbers + test

### PR DESCRIPTION
**Pull request by dawsoneliasen**
This PR updates the covariance formula to be include the complex conjugate ``std::conj()``. `n_samples` must also be cast to double so that covariance matrix can be calculated for complex numbers. A test is added for calculating the covariance matrix for complex samples.

### CODE REVIEW

File: tests/covariance_matrix_04.cc [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  mh_sampler.sample ({ComponentType(0, 0), ComponentType(1, 1)},
+                     &log_likelihood,
+                     &perturb,
+                     8);
+
+  // At this point, we have sampled two sets: first starts from 1 and
+  // goes to 8, second - starts from 2 and goes to 9. Sample variance
+  // for both vector is 6 as covariance too. Output whatever we got:
+  std::cout << covariance_matrix.get()(0,0) << std::endl;
```
Comment (bangerth) [inline tests/covariance_matrix_04.cc:66]: This used to be true for the _02 test, but apparently you now get 48+0i. Have you verified that 48 is the correct answer? If so, update the comment here.

**Comment by dawsoneliasen** [comment]
@bangerth PTAL. Replaced the test case with 3 more simple cases (simple variance). Note that the variance of a sample of complex numbers should actually be equal to the sum of the variances of the real and imaginary components - ie the variance of a complex number is a real number, not a complex number (according to [Wikipedia](https://en.wikipedia.org/wiki/Complex_random_variable#Variance_and_pseudo-variance)). Let's talk about the covariance case tomorrow.

File: tests/covariance_matrix_05.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Check the CovarianceMatrix consumer with complex numbers.
+// This test generates a simple samples that only have one component
+// (so we are really only calculating variance, not covariance) which
+// is a complex number with no real component.
```
Comment (bangerth) [inline tests/covariance_matrix_05.cc:20]: This is something that is at least not obvious to me. Could you add a line or two to the comment here that explains what samples you use, and why the variance for these should really be an imaginary number?

File: tests/covariance_matrix_05.output [hunk 3]
```diff
@@ -0,0 +1 @@
+(6,0)
```
Comment (bangerth) [inline tests/covariance_matrix_05.output:1]: I'll note that what you have here is really a *real* number, not an imaginary number.

File: tests/covariance_matrix_06.output [hunk 4]
```diff
@@ -0,0 +1 @@
+(12,0)
```
Comment (bangerth) [inline tests/covariance_matrix_06.output:1]: But this is not matching what you say in the comment at the top of the .cc file.

File: tests/covariance_matrix_05.cc (hunk 2 above)
Comment (dawsoneliasen) [reply tests/covariance_matrix_05.cc]: I agree that it is confusing. I'm using the word "component" to mean two different things. I've update the comments at the top of each test. But the variance should in fact be a real number, not an imaginary number.

**Pull request merged**